iptables: Add support for IPv6, make IPv4 optional.

jrajahalme opened the pull request:
Use option.Config.EnableIPv4 and option.Config.EnableIPv6 to figure
out whether to install IPv4 rules (with `iptables`) and/or IPv6 rules
(with `ip6tables`).

Xfrm NOTRACK rules are moved to a Cilium chain in the raw PREROUTING
table. Otherwise there are no functional changes in this commit,
assuming that IPv4 is enabled, since none of the Cilium chains specify
the new 'ipv6' option.

These changes are needed for the forthcoming TPROXY support.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7346)
<!-- Reviewable:end -->

jrfastab commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -202,171 +242,173 @@ func InstallRules(ifName string) error {
+		// they are later dropped in the FORWARD chain. The issue has been
+		// resolved in #52569 and will be fixed in k8s >= 1.8. The following is
+		// a workaround for earlier Kubernetes versions.
 		//
```
Just curious if anyone know if k8s 1.8+ has the fix or not?

jrajahalme replied:
@aanm Do you know?

aanm replied:
the commit is available in >=v1.9 https://github.com/kubernetes/kubernetes/commit/2f622b2a2858eea3b74a9f83a368197719b06a6f

@tgraf maybe we can remove this since we are dropping support for 1.8? Does it hurt having this just in case there's a regression in k8s?

tgraf replied:
I'm fine to remove it. The fewer rules the better.

jrfastab commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -326,6 +326,66 @@ func InstallRules(ifName string) error {
+				"-A", ciliumPostNatChain,
+				"-m", "mark", "--mark", matchFromIPSecDecrypt, // Don't match ipsec traffic
+				"-m", "comment", "--comment", "exclude decrypt from masquerade",
+				"-j", "RETURN"}, false); err != nil {
```
I've been considering adding ACCEPT rules for marked packets here. Even if we have rules in other tables/chains I don't believe we want them matching. My recollection is RETURN will continue processing later chains which may not be ideal. However, this code snippet here is a valid transformation of existing code intent so if we want to change it an easy follow up patch works.

jrajahalme replied:
Changed the IPsec exclusion rules to ACCEPT, kept the non-cilium_host exclusion rule as RETURN as it is possible that the packet may need to hit some other rules in the same table.

jrfastab reviewed (approved):
LGTM. CI tests are coming soon for ipsec.

jrajahalme committed:
iptables: Use RETURN targets to exclude traffic from masquarade rules.

Masquarade rules must exclude traffic for other than cilium_host
interface as well as IPsec traffic. TPROXY support (to be added later)
adds all proxy return traffic to these exclusions. Repeating all these
exlusions in each rule is quite verbose, so use RETURN targets in the
Cilium PREROUTING table to exclude traffic from the later masquarade
rules instead.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
iptables: Add support for IPv6, make IPv4 optional.

Use option.Config.EnableIPv4 and option.Config.EnableIPv6 to figure
out whether to install IPv4 rules (with `iptables`) and/or IPv6 rules
(with `ip6tables`).

Xfrm NOTRACK rules are moved to a Cilium chain in the raw PREROUTING
table. Otherwise there are no functional changes in this commit,
assuming that IPv4 is enabled, since none of the Cilium chains specify
the new 'ipv6' option.

These changes are needed for the forthcoming TPROXY support.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
